### PR TITLE
Fix large exports failing

### LINKED ISSUE #5375: Export of large amount of packages will result in errors
**Version**
- Katello 4.12
- pulpcore 3.39.11
- pulp_rpm 3.23.3
- pulp_deb 3.0.2

**Describe the bug**
Exporting data (in this case Content Views in Katello) with large amount of packages will fail:

`Error: sending query and params failed: number of parameters must be between 0 and 65535`

There are actually at least two different issues both of which will throw the same error:

1. If the Content View has more than 65535 packages the export task will fail (usually pretty quickly).
2. If the Content View has less than 65535 but more than around half of that (tested with ~40k packages) the task will actually succeed but fail in writing the data with the same error. This will take long since the export task is actually completed.

Debian and RPM content are both affected by this.

**To Reproduce**

- Create a Content View with more than 65535 packages (or with more than half that but lower than 65535)
- Export the Content View via:
`hammer content-export complete version --chunk-size-gb=40 --content-view="CV_NAME" --organization-id=1 --version=1.0`

**Expected behavior**
The export to be successful.

**Additional context**
We suspect that this has something to do with the upgrade to Django 4.2 or more specifically with the upgrade from psycopg2 to psycopg3. We found an issue that parallels our findings https://github.com/psycopg/psycopg/issues/620

We can also confirm that on older systems that still use older pulpcore versions (pre django 4.2) the export still works on these types of Content Views.

For more context here are the pulp tasks:

1. For Content Views with more than 65535 packages
```
---
pulp_tasks:
- pulp_href: "/pulp/api/v3/tasks/018f767f-1b93-72cb-aa23-32c9c2650ef9/"
  pulp_created: '2024-05-14T09:46:32.725+00:00'
  state: failed
  name: pulpcore.app.tasks.export.pulp_export
  logging_cid: f34cbfee-3afb-4f6a-aff6-83b83e586f0a
  created_by: "/pulp/api/v3/users/1/"
  started_at: '2024-05-14T09:46:33.085+00:00'
  finished_at: '2024-05-14T09:46:55.152+00:00'
  error:
    traceback: |2
        File "/usr/lib/python3.11/site-packages/pulpcore/tasking/tasks.py", line 61, in _execute_task
          result = func(*args, **kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.11/site-packages/pulpcore/app/tasks/export.py", line 417, in pulp_export
          _do_export(pulp_exporter, tar, the_export)
        File "/usr/lib/python3.11/site-packages/pulpcore/app/tasks/export.py", line 512, in _do_export
          export_artifacts(the_export, Artifact.objects.filter(pk__in=artifact_pks))
        File "/usr/lib/python3.11/site-packages/pulpcore/app/importexport.py", line 133, in export_artifacts
          for artifact in pb.iter(artifacts.only("file").iterator()):
        File "/usr/lib/python3.11/site-packages/pulpcore/app/models/progress.py", line 296, in iter
          for x in iter:
        File "/usr/lib/python3.11/site-packages/django/db/models/query.py", line 516, in _iterator
          yield from iterable
        File "/usr/lib/python3.11/site-packages/django/db/models/query.py", line 91, in __iter__
          results = compiler.execute_sql(
                    ^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.11/site-packages/django/db/models/sql/compiler.py", line 1562, in execute_sql
          cursor.execute(sql, params)
        File "/usr/lib/python3.11/site-packages/django/db/backends/utils.py", line 67, in execute
          return self._execute_with_wrappers(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.11/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
          return executor(sql, params, many, context)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.11/site-packages/django/db/backends/utils.py", line 84, in _execute
          with self.db.wrap_database_errors:
        File "/usr/lib/python3.11/site-packages/django/db/utils.py", line 91, in __exit__
          raise dj_exc_value.with_traceback(traceback) from exc_value
        File "/usr/lib/python3.11/site-packages/django/db/backends/utils.py", line 89, in _execute
          return self.cursor.execute(sql, params)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.11/site-packages/psycopg/server_cursor.py", line 294, in execute
          raise ex.with_traceback(None)
    description: 'sending query and params failed: number of parameters must be between
      0 and 65535'
  worker: "/pulp/api/v3/workers/018f73c9-a49a-7797-a12a-836eb2e6a13f/"
  child_tasks: []
  progress_reports:
  - message: Exporting Artifacts
    code: export.artifacts
    state: failed
    total: 66040
    done: 0
  created_resources:
  - "/pulp/api/v3/exporters/core/pulp/018f767f-19d8-7c16-9aa4-be0917d2c868/exports/018f767f-1f26-75cc-921f-b43b68856d32/"
  reserved_resources_record:
  - "/pulp/api/v3/exporters/core/pulp/018f767f-19d8-7c16-9aa4-be0917d2c868/"
  - shared:/pulp/api/v3/repositories/rpm/rpm/018e4331-54d9-7791-b6cf-df2913e135b7/
  - shared:/pulp/api/v3/repositories/rpm/rpm/018e4331-3e03-756e-b478-d7118b396463/
  - shared:/pulp/api/v3/repositories/rpm/rpm/018e4331-81e1-723a-bc88-54750f778760/
  - shared:/pulp/api/v3/repositories/rpm/rpm/018e4331-6a8a-77d2-9bcc-7fea35a5e6fb/
  - shared:/pulp/api/v3/domains/018e42cc-5411-73e1-932c-cc158854d906/
task_groups: []
poll_attempts:
  total: 15
  failed: 1
```

2. For Content Views with around 40k packages:
```
---
pulp_tasks:
- pulp_href: "/pulp/api/v3/tasks/018f768b-022a-764f-a0e1-7ee4e6256654/"
  pulp_created: '2024-05-14T09:59:32.650+00:00'
  state: failed
  name: pulpcore.app.tasks.export.pulp_export
  logging_cid: e0774285-1b15-4539-a9dc-30f2eb8f1764
  created_by: "/pulp/api/v3/users/1/"
  started_at: '2024-05-14T09:59:32.861+00:00'
  finished_at: '2024-05-14T10:27:41.082+00:00'
  error:
    traceback: |2
        File "/usr/lib/python3.11/site-packages/pulpcore/tasking/tasks.py", line 61, in _execute_task
          result = func(*args, **kwargs)
                   ^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.11/site-packages/pulpcore/app/tasks/export.py", line 417, in pulp_export
          _do_export(pulp_exporter, tar, the_export)
        File "/usr/lib/python3.11/site-packages/pulpcore/app/tasks/export.py", line 517, in _do_export
          export_content(the_export, version)
        File "/usr/lib/python3.11/site-packages/pulpcore/app/importexport.py", line 185, in export_content
          _write_export(export.tarfile, resource, dest_dir)
        File "/usr/lib/python3.11/site-packages/pulpcore/app/importexport.py", line 59, in _write_export
          for item in resource.queryset.iterator(chunk_size=EXPORT_BATCH_SIZE):
        File "/usr/lib/python3.11/site-packages/django/db/models/query.py", line 516, in _iterator
          yield from iterable
        File "/usr/lib/python3.11/site-packages/django/db/models/query.py", line 91, in __iter__
          results = compiler.execute_sql(
                    ^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.11/site-packages/django/db/models/sql/compiler.py", line 1562, in execute_sql
          cursor.execute(sql, params)
        File "/usr/lib/python3.11/site-packages/django/db/backends/utils.py", line 67, in execute
          return self._execute_with_wrappers(
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.11/site-packages/django/db/backends/utils.py", line 80, in _execute_with_wrappers
          return executor(sql, params, many, context)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.11/site-packages/django/db/backends/utils.py", line 84, in _execute
          with self.db.wrap_database_errors:
        File "/usr/lib/python3.11/site-packages/django/db/utils.py", line 91, in __exit__
          raise dj_exc_value.with_traceback(traceback) from exc_value
        File "/usr/lib/python3.11/site-packages/django/db/backends/utils.py", line 89, in _execute
          return self.cursor.execute(sql, params)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        File "/usr/lib/python3.11/site-packages/psycopg/server_cursor.py", line 294, in execute
          raise ex.with_traceback(None)
    description: 'sending query and params failed: number of parameters must be between
      0 and 65535'
  worker: "/pulp/api/v3/workers/018f73c9-accc-7c2e-ba0d-0c381a66be93/"
  child_tasks: []
  progress_reports:
  - message: Exporting Artifacts
    code: export.artifacts
    state: completed
    total: 48052
    done: 48052
  created_resources:
  - "/pulp/api/v3/exporters/core/pulp/018f768b-0112-7365-bfd8-e30a0cd8eeb6/exports/018f768b-0448-7bcd-abd8-052060d1d70d/"
  reserved_resources_record:
  - "/pulp/api/v3/exporters/core/pulp/018f768b-0112-7365-bfd8-e30a0cd8eeb6/"
  - shared:/pulp/api/v3/repositories/rpm/rpm/018e4331-81e1-723a-bc88-54750f778760/
  - shared:/pulp/api/v3/domains/018e42cc-5411-73e1-932c-cc158854d906/
task_groups: []
poll_attempts:
  total: 123
  failed: 1
```

### PR DESCRIPTION
**Pull request by hstct**
closes #5375

### CODE REVIEW

**Comment by hstct** [comment]
This is a still in progress work. It does not account for cases with other storage backends. And the solution for the string operation of the artifact paths is still hackish.

However it did successfully export a large RPM content view with more than 65k packages which previously failed.

File: pulpcore/app/importexport.py [hunk 1]
```diff
@@ -102,23 +103,26 @@ def export_versions(export, version_info):
...
+    artifacts = Artifact.objects.filter(pk__in=artifact_pks)
+    artifact_files = Artifact.objects.filter(pk__in=artifact_pks).values_list("file", flat=True)
     data = dict(message="Exporting Artifacts", code="export.artifacts", total=len(artifacts))
     with ProgressReport(**data) as pb:
         pb.BATCH_INTERVAL = 5000
 
         if settings.DEFAULT_FILE_STORAGE != "pulpcore.app.models.storage.FileSystem":
             with tempfile.TemporaryDirectory(dir=".") as temp_dir:
+                # TODO: The following line is still affected by the >65k artifacts issue!!!
                 for artifact in pb.iter(artifacts.only("file").iterator()):
```
Comment (quba42) [inline pulpcore/app/importexport.py:126]: This line has not been fixed yet. Unfortunately we have no way of testing this branch of the if tree, so we are unsure how to proceed with it.

File: pulpcore/app/importexport.py [hunk 2]
```diff
@@ -130,8 +134,11 @@ def export_artifacts(export, artifacts):
                         artifact.file.close()
                         export.tarfile.add(temp_file.name, artifact.file.name)
         else:
-            for artifact in pb.iter(artifacts.only("file").iterator()):
-                export.tarfile.add(artifact.file.path, artifact.file.name)
+            for artifact_file in pb.iter(artifact_files):
+                # TODO: Hard coding the base path in the following line is not robust
+                # There is also an assumption that artifact.file == artifact.file.name
+                artifact_file_path = os.path.join("/var/lib/pulp/media/", artifact_file)
+                export.tarfile.add(artifact_file_path, artifact_file)
```
Comment (quba42) [inline pulpcore/app/importexport.py:141]: The essence of the fix is to use `.values_list("file", flat=True)`, however that means we never get instantiated `Aritfact` objects with instantiated `ArtifactFileField` objects on their `file` field. Does anyone know of a robust way to nevertheless access `file.name` and `file.path` to replace my hackish workaround above?

File: pulpcore/app/importexport.py (hunk 1 above)
Comment (quba42) [reply pulpcore/app/importexport.py]: Fixed using a batching solution.

File: pulpcore/app/importexport.py (hunk 2 above)
Comment (quba42) [reply pulpcore/app/importexport.py]: Fixed using a batching solution.

File: pulpcore/app/importexport.py [hunk 3]
```diff
@@ -54,22 +55,17 @@ def process_batch(batch):
...
-                    process_batch(batch)
-                    batch.clear()
-                    needs_comma = True
-
-            if batch:
                 process_batch(batch)
+                needs_comma = True
 
             temp_file.write("]")
         else:
```
Comment (quba42) [inline pulpcore/app/importexport.py:71]: I have now idea how to end up in this branch, but I don't think anything needs fixing there...

File: pulpcore/app/importexport.py [hunk 4]
```diff
@@ -102,36 +98,48 @@ def export_versions(export, version_info):
...
     """
-    data = dict(message="Exporting Artifacts", code="export.artifacts", total=len(artifacts))
+    artifacts = Artifact.objects.filter(pk__in=artifact_pks)
+    data = dict(message="Exporting Artifacts", code="export.artifacts", total=len(artifact_pks))
     with ProgressReport(**data) as pb:
         pb.BATCH_INTERVAL = 5000
 
+        batch = True
+        offset = 0
         if settings.DEFAULT_FILE_STORAGE != "pulpcore.app.models.storage.FileSystem":
```
Comment (quba42) [inline pulpcore/app/importexport.py:119]: This branch is untested by us, but the implementation is very similar to the other branch.

File: pulpcore/app/importexport.py [hunk 5]
```diff
@@ -54,21 +55,18 @@ def process_batch(batch):
...
-                    needs_comma = False
-
-                if len(batch) >= EXPORT_BATCH_SIZE:
-                    process_batch(batch)
-                    batch.clear()
-                    needs_comma = True
+            offset = 0
+            first_loop = True
+            while True:
+                batch = list(resource.queryset[offset : offset + EXPORT_BATCH_SIZE])
```
Comment (dralley) [inline pulpcore/app/importexport.py:61]: I believe this queryset isn't necessarily time-stable or ordering-stable.  That is, by splitting one query into multiple queries and indexing into it, you are not necessarily getting the same objects in the same order as if it was all in one query.  At minimum you'd probably need to do an ```.order_by("pk")``` or something but that only fixes the ordering issue

Comment (dralley) [reply pulpcore/app/importexport.py:61]: I think what we might want to do instead is batch the list of PKs (produced using `.values_list()`, then generate the querysets later at batch time

Comment (quba42) [reply pulpcore/app/importexport.py]: I pushed some changes that convert to batching over a list of pks and creating a query for each batch.
I am not sure on the following points:

1. Should we still add an `.order_by("pk")` to the individual batch queries or is it not necessary to preserve order within a batch?
2. Do we need to do the same thing for the batching over the artifacts within the `export_artifacts` function?

Comment (quba42) [reply pulpcore/app/importexport.py]: I implemented (2.) in a [separate commit on this PR](https://github.com/pulp/pulpcore/pull/5388/commits/ea0e8b462d1843cff7dfb2643efbf0f07286c60e) feel free to review it and then we can decide to keep it or not.

Comment (dralley) [reply pulpcore/app/importexport.py]: It shouldn't be necessary to preserve order within a batch

Comment (dralley) [reply pulpcore/app/importexport.py]: Yes please, just one commit.  It's much easier to backport (our automation has trouble with backporting more than one commit)

**Review by dralley — APPROVED** [review]
This looks reasonable!

We need some more extensive import/export testing, it's just difficult to set up.

Let's squash the commits and get one additional review

File: pulpcore/app/importexport.py (hunk 5 above)
Comment (quba42) [reply pulpcore/app/importexport.py]: Done.

File: pulpcore/app/importexport.py [hunk 6]
```diff
@@ -102,39 +99,48 @@ def export_versions(export, version_info):
...
-                        #       artifact.file.read() directly to the tarfile with tarfile.addfile()
-                        temp_file.write(artifact.file.read())
-                        temp_file.flush()
-                        artifact.file.close()
-                        export.tarfile.add(temp_file.name, artifact.file.name)
+                for offset in range(0, len(artifact_pks), EXPORT_BATCH_SIZE):
+                    batch = artifact_pks[offset : offset + EXPORT_BATCH_SIZE]
+                    batch_qs = Artifact.objects.filter(pk__in=batch).only("file")
+
+                    for artifact in pb.iter(batch_qs):
```
Comment (ipanova) [inline pulpcore/app/importexport.py:123]: you could use iterator() here as well

Comment (ipanova) [reply pulpcore/app/importexport.py]: https://docs.djangoproject.com/en/dev/ref/models/querysets/#django.db.models.query.QuerySet.iterator

**Review by ipanova — APPROVED** [review]

File: pulpcore/app/importexport.py (hunk 6 above)
Comment (quba42) [reply pulpcore/app/importexport.py]: Reading that, `.iterator()` sounds like the right choice for batch query sets. Thanks for the link.

Comment (dralley) [reply pulpcore/app/importexport.py]: I doubt it actually mattered in this instance as the default batch size is 2000, and we're already doing external batching... with a batch size of 2000

But it doesn't hurt anything so no need to revert.

**Review by dralley — APPROVED** [review]

**Pull request merged**